### PR TITLE
feat: custom scope

### PR DESCRIPTION
**Pull request by IceHe**
## Summary

- The developers can use custom `scope` value.
- LogtoClient can correct the value of `scope` from parameters.

### CODE REVIEW

File: packages/client/src/index.ts [hunk 1]
```diff
@@ -22,6 +23,7 @@ export interface ConfigParameters {
   domain: string;
   clientId: string;
   storage?: ClientStorage;
+  scope: string | string[];
```
Comment (wangsijie) [inline packages/client/src/index.ts:26]: can be optional

File: packages/client/src/index.ts [hunk 2]
```diff
@@ -35,27 +37,53 @@ export const appendSlashIfNeeded = (url: string): string => {
...
+    const { clientId, storage, scope } = config;
     this.clientId = clientId;
+    this.scope = LogtoClient.correctScope(scope);
     this.oidcConfiguration = oidcConfiguration;
     this.storage = storage ?? new LocalStorage();
     this.sessionManager = new SessionManager(this.storage);
     this.createTokenSetFromCache();
   }
 
+  static correctScope(originalScope: string | string[]): string {
```
Comment (wangsijie) [inline packages/client/src/index.ts:54]: put it into utils? no need to be public

File: packages/client/src/index.ts [hunk 3]
```diff
@@ -35,27 +37,53 @@ export const appendSlashIfNeeded = (url: string): string => {
...
+      }
+
+      return OPENID + ' ' + originalScope.map((s) => s.trim()).join(' ');
+    }
+
+    if (originalScope.trim().length === 0) {
+      return OPENID;
+    }
+
+    if (originalScope.split(' ').includes(OPENID)) {
```
Comment (wangsijie) [inline packages/client/src/index.ts:71]: this `includes` code is seen twice,
how about turning string value into `array` and then check default scopes,
also, `offline_access` is necessary as well

**Review by wangsijie — CHANGES_REQUESTED** [review]

File: packages/client/src/index.ts [hunk 4]
```diff
@@ -35,27 +37,53 @@ export const appendSlashIfNeeded = (url: string): string => {
...
-    const client = new LogtoClient(config, await discover(`https://${config.domain}`));
-    return client;
+    return new LogtoClient(config, await discover(`https://${config.domain}`));
   }
 
   public async loginWithRedirect(redirectUri: string) {
     const { url, codeVerifier } = await getLoginUrlAndCodeVerifier(
       this.oidcConfiguration.authorization_endpoint,
       this.clientId,
+      this.scope,
```
Comment (simeng-li) [inline packages/client/src/index.ts:86]: there are too many input attributes here, better make it a single payload map

File: packages/client/src/index.ts (hunk 2 above)
Comment (simeng-li) [reply packages/client/src/index.ts:54]: could align with auth0 only accepts string with empty space split the scopes

File: packages/client/src/index.test.ts [hunk 5]
```diff
@@ -16,9 +16,11 @@ jest.mock('./verify-id-token');
 const DOMAIN = 'logto.dev';
 const BASE_URL = `https://${DOMAIN}`;
 const CLIENT_ID = 'client1';
+const OPENID = 'openid';
+const SCOPE = OPENID;
```
Comment (simeng-li) [inline packages/client/src/index.test.ts:20]: feels like no need to define a const SCOPE = OPENID here.  We could define the value whenever it is being used. i.e. new LogtoClient
as scope should be optional param.

1. be default no need to pass in Scope to all the current existing tests. 
2. Create a test does not pass in Scope as check the `client.scope` contains OPENID
3. Create a test pass scope = OPENID, check the `client.scope` contains only one single OPENID
4. Create a test pass scope = `${OPENID} email`, check the `client.scope` contains OPENID and email

should be good enough

File: packages/client/src/index.ts (hunk 2 above)
Comment (simeng-li) [reply packages/client/src/index.ts:54]: can we could have the trim check on the original scope input param

Comment (IceHe) [reply packages/client/src/index.ts:54]: i prefer `string | string[]`

**Review by wangsijie — CHANGES_REQUESTED** [review]
`tokenSetCacheKey` should include `scope`

**Review by wangsijie — APPROVED** [review]

File: packages/client/src/index.ts [hunk 6]
```diff
@@ -149,7 +153,7 @@ export default class LogtoClient {
   }
 
   private get tokenSetCacheKey() {
-    return `${TOKEN_SET_CACHE_KEY}::${this.oidcConfiguration.issuer}::${this.clientId}`;
+    return `${TOKEN_SET_CACHE_KEY}::${this.oidcConfiguration.issuer}::${this.clientId}::${this.scope}`;
```
Comment (simeng-li) [inline packages/client/src/index.ts:156]: @IceHe create a issue replace spaces in scope as it is being used as party of the key.

File: packages/client/src/utils.ts [hunk 7]
```diff
@@ -57,3 +57,19 @@ export const nowRoundToSec = () => Math.floor(Date.now() / 1000);
...
+ */
+export const generateScope = (originalScope?: string | string[]): string => {
+  const originalScopeValues =
+    originalScope === undefined
+      ? []
+      : Array.isArray(originalScope)
+      ? originalScope
+      : originalScope.split(' ');
+  const nonEmptyScopeValues = originalScopeValues.map((s) => s.trim()).filter((s) => s.length > 0);
+  const uniqueScopeValues = new Set(['openid', 'offline_access', ...nonEmptyScopeValues]);
```
Comment (simeng-li) [inline packages/client/src/utils.ts:73]: better create a const for default scopes

**Pull request merged**